environments/centos7-base/main.py

```diff
@@ -0,0 +1,204 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+import stat
+import json
+import urllib2
+
+from datetime import datetime
+from subprocess import call
+from string import Template
+
+def error(message):
+    log(Template('ERR: $message').substitute(message=message))
+    sys.stdout.flush()
+
+def warn(message):
+    log(Template('WARN: $message').substitute(message=message))
+    sys.stdout.flush()
+
+def log(message):
+    print(Template('[main] $message').substitute(message=message))
+    sys.stdout.flush()
+
+# This function will attempt to read any results file, add our metadata to it,
+# write it to disk, send everything to S3, and exit the program.
+#
+# If succeeded is false, we'll assume things failed badly enough that we shouldn't
+# even try to read /grade/results/results.json.
+def finish(succeeded, info):
+    if info['job_id'] is None:
+        # Someone screwed up bad; without a job ID, we can't upload archives to
+        # S3 or notify the webhook that we've failed
+        error('job_id was not specified, so it\'s impossible to upload an archive')
+        sys.exit(1)
+
+    data = None
+    if succeeded and os.path.isfile('/grade/results/results.json'):
```

Comment:
I never knew TOCTTOU was a thing! Thanks for teaching about that.

Re: behavior if we fail to parse the JSON that our "client" has written to disk, I'm not really sure what the appropriate thing to do is. Should we mark the run as failed and add a message?